### DISCUSSION TITLE
Linux kernel hang at  boot....

### DISCUSSION THREAD

**Discussion by cacheman** [ipxe/ipxe#1260] [Q&A]
Hi.....

I have a system which uses a static compiled (i.e. no initrd) Linux 6.x EFI kernel which I boot using iPXE using a menu entry like this:

`:TestLinux`
`kernel http://aaa.bbb.ccc.ddd/bzImage root=/dev/nfs nfsroot=aaa.bbb.ccc.ddd:/export/PC,vers=3,rsize=8192,wsize=8192,actimeo=9999 ip=dhcp nofb nomodeset nohugeiomap vga=0x317`
`boot || goto failed`
`goto start`

This has worked fine on all hardware we have (desktops, laptops etc.) until now, but recently we received a number of new Dell Latitude 5550 laptops, and I can't get the above linux kernel to boot.....all I get is the following:

`http://aaa.bbb.ccc.ddd/bzImage... ok`
`boot`

.....and it just sits at 'boot' with no output. Once it gets to this point the only thing I can do is power cycle the laptop....Ctrl-Alt-Del doesn't work.

The kernel that has been working fine for quite a while (and still works on all of our hardware except this model of laptop) is 6.2.1. I have tried building the latest 6.10 kernel, but I still get the same problem with this laptop model....I have even tried an old 5.x kernel, but still the same problem.

I have also tried enabling early printk messages (using kernel command line options earlyprintk and loglevel), but I get no output of any kind from the kernel....which suggests it's not even beginning to boot....

....and I have tried building the latest iPXE git clone, but no joy....still hangs at same place with no output.

I have run an `imgstat` command after the `kernel` iPXE command to see what iPXE thinks it loaded and it looks fine to me:

`bzImage: 11170816 bytes [EFI] [SELECTED] "root=/dev/nfs nfsroot=aaa.bbb.ccc.ddd:/export/PC,vers=3,rsize=8192,wsize=8192,actimeo=9999 ip=dhcp nofb nomodeset nohugeiomap vga=0x317"`

Unfortunately having no error message or any message of any kind makes it really hard to debug....

I was just wondering if there are any debug options I could use in iPXE to have the boot command and whatever follows display some kind of debug information which might help me track down the cause of this.....

....or perhaps someone reading this might have seen the same or a similar problem and might be able to shed some light on what could be causing this....?

Thanks in advance for any help.

Cacheman.

**Comment by NiKiZe** [comment]
In EFI, iPXE calls the firmware to boot the EFI binary (your kernel) there is not much that can be debugged.

You could verify this by placing the kernel as `bootx64.efi` in the correct directory on a fat32 formatted USB stick.

You might also want to verify that your other working cases is using EFI as well and not pcbios.

**Comment by cacheman** [comment]
Hi....

Thanks for the quick reply. All of the devices that currently do work are using EFI.....

Cacheman.

**Comment by cacheman** [comment]
Hi....

Just wondering.....could it be something not compiled correctly in my kernel? An kernel build option (EFI or otherwise) I am missing and that this latest Dell laptop needs?

The reason I ask this is that I can iPXE boot a RedHat EL9 KickStart installer via EFI on the same laptop, and I do get to the RH installer screen....my menu entry for this is:

`:KS_RHEL9`
`echo Booting RHEL9`
`initrd http://aaa.bbb.ccc.ddd/RHEL9/OS/x86_64/dist/images/pxeboot/initrd.img`
`chain http://aaa.bbb.ccc.ddd/RHEL9/OS/x86_64/dist/images/pxeboot/vmlinuz usb-storage.delay_use=10 inst.selinux=0 inst.text ksdevice=bootif noipv6 load_ramdisk=1 initrd=initrd.img network inst.ks=http://aaa.bbb.ccc.ddd/RHEL9/cgi-bin/ks-rhel9-x86_64-initial.cfg os=RHEL9 BOOTIF=01-${net0/mac}`

Would be grateful for any pointers.....

TIA

Chris.

**Reply by NiKiZe** [reply]
Missing cpu instructions and you kernel targeting a different cpu, hard to tell.
I don't think this is the place to debug a kernel boot issue, unless you can say it only happens when booting thru iPXE,

Have you tried any other bootloader, or naming the kernel bootx64.efi and placed on usb as explained earlier?

**Comment by cacheman** [comment]
Hi....

Thanks for the reply....

Apologies....I knew when I posted this there was a good chance that this forum might not be the right place for this, but in the absence of an error message to guide me, I had to start somewhere and my first thought was that it might be a problem with iPXE booting the kernel since that was where it hung.

I have since tried what you suggested renaming the kernel to bootx64.efi and putting it on a USB stick....and it displays the following message:

EFI stub: WARNING: [Firmware Bug]: LoadOptions is an EFI_LOAD_OPTION descriptor
EFI stub: WARNING: [Firmware Bug]: Using OptionalData as a workaround

.....which I would assume this means is not an iPXE issue. As before, after this I get no output. 

I have confirmed that this bootx64.efi on USB does work properly on any older hardware that was already working with iPXE.

I have also configured a USB stick with Grub and on pre-existing already working hardware this works fine, but on this new laptop, same a before.....laptop just sits there with no output (not even the EFI stub warning above).....

Based on some googl'ing it sounds like this could be a framebuffer issue but I'm not sure how to debug/prove this.....

Maybe this laptop is too recent or has some very recent quirk that isn't yet supported in the kernel....

Anyway, I'll try posting this to a kernel forum and see how far I get there. 

In the meantime, if I do find a solution, I'll try to come back here to update. 

....and of course, even though this seems to not be an iPXE issue, any suggestions on things to try from anyone here would be most welcome.....

Thanks for the help.

Cacheman

**Reply by NiKiZe** [reply]
Try a clean default build of the kernel and see if that works, if it does, start checking for config differences, EFI framebuffer might be a good bet, but  it might be better to have no assumptions here.

I would be looking for a hang, rather than a "no output" issue.

The warnings you are seeing is probably irrelevant, and just about no options being given to the kernel in that scenario.

In terms of testing configs, I would create a git repo, put a good config in it, then add your current, and use diff tools, also switching half of the config options and then the other half, and then plausibly using the git bisect feature. (I love abusing git features in different ways, sorry)
